subprojects/android-test/toast-rule/src/main/kotlin/com/avito/android/rule/internal/ProxyToastChecks.kt

```diff
@@ -0,0 +1,110 @@
+package com.avito.android.rule.internal
+
+import android.content.Context
+import android.view.View
+import android.widget.Toast
+import androidx.annotation.StringRes
+import androidx.test.espresso.matcher.ViewMatchers
+import com.avito.android.rule.ToastChecks
+import com.avito.android.test.waitFor
+import org.hamcrest.Description
+import org.hamcrest.Matcher
+import org.hamcrest.MatcherAssert
+import org.hamcrest.Matchers
+import org.hamcrest.Matchers.empty
+import org.hamcrest.Matchers.not
+import org.hamcrest.TypeSafeMatcher
+
+internal class ProxyToastChecks(
+    private val proxyToast: MockProxyToast,
+    private val textAccessor: ToastTextAccessor,
+) : ToastChecks {
+
+    private val longToastDuration = 3500L
+
+    override fun toastDisplayedWithText(text: String) {
+        internalToastDisplayedWithText(text)
+    }
+
+    override fun toastDisplayedWithText(@StringRes resId: Int) {
+        val toast = firstToastOrThrow()
+        val context = toast.getFieldByReflection("mContext") as Context
+        val text = context.getString(resId)
+
+        internalToastDisplayedWithText(text)
+    }
+
+    private fun internalToastDisplayedWithText(text: String) {
+        waitFor(
+            allowedExceptions = setOf(AssertionError::class.java),
+            timeoutMs = longToastDuration
+        ) {
+            val texts = proxyToast.shownToasts.mapNotNull(textAccessor::text)
+
+            MatcherAssert.assertThat(
+                "Toast has shown with text that matches",
+                texts,
+                Matchers.hasItem(text)
+            )
+        }
+    }
+
+    @Suppress("OverridingDeprecatedMember")
+    override fun toastDisplayedWithText(matcher: Matcher<View>) =
+        toastDisplayedWithView(matcher)
+
+    override fun toastDisplayedWithView(matcher: Matcher<View>) {
+        waitFor(
+            allowedExceptions = setOf(AssertionError::class.java),
+            timeoutMs = longToastDuration
+        ) {
+            MatcherAssert.assertThat(
+                "Toast has shown with text that matches",
+                proxyToast.shownToasts,
+                Matchers.hasItem(withView(matcher))
+            )
+        }
+    }
+
+    override fun toastNotDisplayed() {
+        MatcherAssert.assertThat(
+            "There are no toasts shown",
+            proxyToast.shownToasts,
+            empty()
+        )
+    }
+
+    private fun firstToastOrThrow(): Toast {
+        return waitFor(
+            allowedExceptions = setOf(AssertionError::class.java),
+            timeoutMs = longToastDuration
+        ) {
+            MatcherAssert.assertThat(
+                "There are no toasts shown",
+                proxyToast.shownToasts,
+                not(empty())
+            )
+            proxyToast.shownToasts.first()
+        }
+    }
+
+    private fun withView(viewMatcher: Matcher<View>): Matcher<Toast> = object : TypeSafeMatcher<Toast>() {
```

Comment:
Why do you think that we don't need a class?

Reply:
Fixed, introduced a class